"multi_state" failed question

Hi thanks for this awesome project! I have a quick question here.

I would like to use the "multi_state" feature but don't know if I config it correctly. this is my tfmigrate_mv_dir1_dir2.hcl file in sandbox env

```
bash-5.0# cat tfmigrate_mv_dir1_dir2.hcl 
migration "multi_state" "mv_dir1_dir2" {
  from_dir = "/work/tmp/dir1"
  to_dir   = "/work/tmp/dir2"
  actions = [
    "mv aws_security_group.foo aws_security_group.foo2",
    "mv aws_security_group.bar aws_security_group.bar2",
  ]
}
```

and this is the current state list
```
bash-5.0# terraform state list
aws_security_group.bar
aws_security_group.foo
```

/work/tmp/dir2 is a empty directory
```
bash-5.0# ls -al /work/tmp/dir2
total 0
drwxr-xr-x    2 root     root            64 Oct  9 02:03 .
drwxr-xr-x    4 root     root           128 Oct  8 14:32 ..
```

What I think multi_state will do for me is
1. rename the resource 
2. cut from the current tfstate file from dir1 and create a new tfstate file in dir2 
But, instead, this is the result. 
```
bash-5.0# TFMIGRATE_LOG=DEBUG tfmigrate plan tfmigrate_mv_dir1_dir2.hcl
2021/10/09 02:05:31 [DEBUG] [main] start: tfmigrate plan tfmigrate_mv_dir1_dir2.hcl
2021/10/09 02:05:31 [DEBUG] [main] tfmigrate version: 0.2.9
2021/10/09 02:05:31 [DEBUG] [command] config: &config.TfmigrateConfig{MigrationDir:".", History:(*history.Config)(nil)}
2021/10/09 02:05:31 [DEBUG] [command] option: &tfmigrate.MigratorOption{ExecPath:"", PlanOut:""}
2021/10/09 02:05:31 [INFO] [runner] load migration file: tfmigrate_mv_dir1_dir2.hcl
2021/10/09 02:05:31 [INFO] [migrator] multi start state migrator plan
2021/10/09 02:05:31 [DEBUG] [executor@/work/tmp/dir1]$ terraform version
2021/10/09 02:05:31 [INFO] [migrator@/work/tmp/dir1] terraform version: 1.0.8
2021/10/09 02:05:31 [INFO] [migrator@/work/tmp/dir1] initialize work dir
2021/10/09 02:05:32 [DEBUG] [executor@/work/tmp/dir1]$ terraform init -input=false -no-color
2021/10/09 02:05:32 [INFO] [migrator@/work/tmp/dir1] switch to remote workspace default
2021/10/09 02:05:33 [DEBUG] [executor@/work/tmp/dir1]$ terraform workspace select default
2021/10/09 02:05:33 [INFO] [migrator@/work/tmp/dir1] get the current remote state
2021/10/09 02:05:33 [DEBUG] [executor@/work/tmp/dir1]$ terraform state pull
2021/10/09 02:05:33 [INFO] [migrator@/work/tmp/dir1] override backend to local
2021/10/09 02:05:33 [INFO] [executor@/work/tmp/dir1] create an override file
2021/10/09 02:05:33 [INFO] [migrator@/work/tmp/dir1] creating local workspace folder in: /work/tmp/dir1/terraform.tfstate.d/default
2021/10/09 02:05:33 [INFO] [executor@/work/tmp/dir1] switch backend to local
2021/10/09 02:05:35 [DEBUG] [executor@/work/tmp/dir1]$ terraform init -input=false -no-color -reconfigure
2021/10/09 02:05:35 [DEBUG] [executor@/work/tmp/dir2]$ terraform version
2021/10/09 02:05:35 [INFO] [migrator@/work/tmp/dir2] terraform version: 1.0.8
2021/10/09 02:05:35 [INFO] [migrator@/work/tmp/dir2] initialize work dir
2021/10/09 02:05:35 [DEBUG] [executor@/work/tmp/dir2]$ terraform init -input=false -no-color
2021/10/09 02:05:35 [INFO] [migrator@/work/tmp/dir2] switch to remote workspace default
2021/10/09 02:05:35 [DEBUG] [executor@/work/tmp/dir2]$ terraform workspace select default
2021/10/09 02:05:35 [INFO] [migrator@/work/tmp/dir2] get the current remote state
2021/10/09 02:05:36 [DEBUG] [executor@/work/tmp/dir2]$ terraform state pull
2021/10/09 02:05:36 [INFO] [migrator@/work/tmp/dir2] override backend to local
2021/10/09 02:05:36 [INFO] [executor@/work/tmp/dir2] create an override file
2021/10/09 02:05:36 [INFO] [migrator@/work/tmp/dir2] creating local workspace folder in: /work/tmp/dir2/terraform.tfstate.d/default
2021/10/09 02:05:36 [INFO] [executor@/work/tmp/dir2] switch backend to local
2021/10/09 02:05:36 [DEBUG] [executor@/work/tmp/dir2]$ terraform init -input=false -no-color -reconfigure
2021/10/09 02:05:36 [INFO] [migrator] compute new states (/work/tmp/dir1 => /work/tmp/dir2)
2021/10/09 02:05:36 [DEBUG] [executor@/work/tmp/dir1]$ terraform state mv -state=/tmp/tmp995693296 -state-out=/tmp/tmp740880783 -backup=/dev/null aws_security_group.foo aws_security_group.foo
2021/10/09 02:05:36 [DEBUG] [executor@/work/tmp/dir2]$ terraform state mv -state=/tmp/tmp848061346 -state-out=/tmp/tmp388647065 -backup=/dev/null aws_security_group.foo aws_security_group.foo2
2021/10/09 02:05:36 [DEBUG] [executor@/work/tmp/dir1]$ terraform state mv -state=/tmp/tmp335183140 -state-out=/tmp/tmp391829555 -backup=/dev/null aws_security_group.bar aws_security_group.bar
2021/10/09 02:05:36 [DEBUG] [executor@/work/tmp/dir2]$ terraform state mv -state=/tmp/tmp514207734 -state-out=/tmp/tmp343693021 -backup=/dev/null aws_security_group.bar aws_security_group.bar2
2021/10/09 02:05:36 [INFO] [migrator@/work/tmp/dir1] check diffs
2021/10/09 02:05:37 [DEBUG] [executor@/work/tmp/dir1]$ terraform plan -state=/tmp/tmp664858264 -out=/tmp/tfplan877530391 -input=false -no-color -detailed-exitcode
2021/10/09 02:05:37 [DEBUG] [executor@/work/tmp/dir1] failed to run command: (*exec.ExitError)(0xc0003063a0)(exit status 2)
2021/10/09 02:05:37 [ERROR] [migrator@/work/tmp/dir1] unexpected diffs
2021/10/09 02:05:37 [INFO] [executor@/work/tmp/dir2] remove the override file
2021/10/09 02:05:37 [INFO] [executor@/work/tmp/dir2] remove the workspace state folder
2021/10/09 02:05:37 [INFO] [executor@/work/tmp/dir2] switch back to remote
2021/10/09 02:05:38 [DEBUG] [executor@/work/tmp/dir2]$ terraform init -input=false -no-color -reconfigure
2021/10/09 02:05:38 [INFO] [executor@/work/tmp/dir1] remove the override file
2021/10/09 02:05:38 [INFO] [executor@/work/tmp/dir1] remove the workspace state folder
2021/10/09 02:05:38 [INFO] [executor@/work/tmp/dir1] switch back to remote
2021/10/09 02:05:40 [DEBUG] [executor@/work/tmp/dir1]$ terraform init -input=false -no-color -reconfigure
terraform plan command returns unexpected diffs: failed to run command (exited 2): terraform plan -state=/tmp/tmp664858264 -out=/tmp/tfplan877530391 -input=false -no-color -detailed-exitcode
stdout:

Terraform used the selected providers to generate the following execution
plan. Resource actions are indicated with the following symbols:
  + create

Terraform will perform the following actions:

  # aws_security_group.bar will be created
  + resource "aws_security_group" "bar" {
      + arn                    = (known after apply)
      + description            = "Managed by Terraform"
      + egress                 = (known after apply)
      + id                     = (known after apply)
      + ingress                = (known after apply)
      + name                   = "bar"
      + name_prefix            = (known after apply)
      + owner_id               = (known after apply)
      + revoke_rules_on_delete = false
      + tags_all               = (known after apply)
      + vpc_id                 = (known after apply)
    }

  # aws_security_group.foo will be created
  + resource "aws_security_group" "foo" {
      + arn                    = (known after apply)
      + description            = "Managed by Terraform"
      + egress                 = (known after apply)
      + id                     = (known after apply)
      + ingress                = (known after apply)
      + name                   = "foo"
      + name_prefix            = (known after apply)
      + owner_id               = (known after apply)
      + revoke_rules_on_delete = false
      + tags_all               = (known after apply)
      + vpc_id                 = (known after apply)
    }

Plan: 2 to add, 0 to change, 0 to destroy.

─────────────────────────────────────────────────────────────────────────────

Saved the plan to: /tmp/tfplan877530391

To perform exactly these actions, run the following command to apply:
    terraform apply "/tmp/tfplan877530391"

stderr:
``` 

Instead rename and move the tfstate it seem to want to create extra resource for me. and now I wondering on which step I might did wrong on multi_state.

Thank for your help and your awesome project once again!

ok @minamijoyo,

I recreate the POC once again and please let me know if I done correctly this time.

this is what I have in dir1
```
bash-5.0# ls -al
total 28
drwxr-xr-x   11 root     root           352 Oct  9 02:15 .
drwxr-xr-x    4 root     root           128 Oct  8 14:32 ..
drwxr-xr-x    4 root     root           128 Oct  9 02:15 .terraform
-rw-r--r--    1 root     root          1077 Oct  7 21:10 .terraform.lock.hcl
-rw-r--r--    1 root     root           107 Oct  8 14:41 .tfmigrate
-rw-r--r--    1 root     root          1114 Oct  9 02:14 config.tf
-rw-r--r--    1 root     root             1 Oct  9 02:14 main.tf
drwxr-xr-x    4 root     root           128 Oct  9 02:12 old
-rw-r--r--    1 root     root           235 Oct  9 02:02 tfmigrate_mv_dir1_dir2.hcl
-rw-r--r--    1 root     root           107 Oct  8 14:47 tfmigrate_test.hcl
-rw-r--r--    1 root     root           174 Oct  8 15:23 tfmigrate_test2.hcl

bash-5.0# cat main.tf 
NULL
bash-5.0#
```

this is what I have in dir2
```
bash-5.0# pwd
/work/tmp/dir2
bash-5.0# ls -al
total 12
drwxr-xr-x    6 root     root           192 Oct  9 02:15 .
drwxr-xr-x    4 root     root           128 Oct  8 14:32 ..
-rw-r--r--    1 root     root          1114 Oct  9 02:12 config.tf
-rw-r--r--    1 root     root           111 Oct  9 02:12 main.tf

bash-5.0# cat main.tf 
resource "aws_security_group" "foo" {
  name = "foo"
}

resource "aws_security_group" "bar" {
  name = "bar"
}
```

here is state list
```
bash-5.0# terraform state list
aws_security_group.bar
aws_security_group.foo
```

here is multi-state running results
```
bash-5.0# TFMIGRATE_LOG=DEBUG tfmigrate plan tfmigrate_mv_dir1_dir2.hcl
2021/10/09 02:33:24 [DEBUG] [main] start: tfmigrate plan tfmigrate_mv_dir1_dir2.hcl
2021/10/09 02:33:24 [DEBUG] [main] tfmigrate version: 0.2.9
2021/10/09 02:33:24 [DEBUG] [command] config: &config.TfmigrateConfig{MigrationDir:".", History:(*history.Config)(nil)}
2021/10/09 02:33:24 [DEBUG] [command] option: &tfmigrate.MigratorOption{ExecPath:"", PlanOut:""}
2021/10/09 02:33:24 [INFO] [runner] load migration file: tfmigrate_mv_dir1_dir2.hcl
2021/10/09 02:33:24 [INFO] [migrator] multi start state migrator plan
2021/10/09 02:33:24 [DEBUG] [executor@/work/tmp/dir1]$ terraform version
2021/10/09 02:33:24 [INFO] [migrator@/work/tmp/dir1] terraform version: 1.0.8
2021/10/09 02:33:24 [INFO] [migrator@/work/tmp/dir1] initialize work dir
2021/10/09 02:33:25 [DEBUG] [executor@/work/tmp/dir1]$ terraform init -input=false -no-color
2021/10/09 02:33:25 [INFO] [migrator@/work/tmp/dir1] switch to remote workspace default
2021/10/09 02:33:26 [DEBUG] [executor@/work/tmp/dir1]$ terraform workspace select default
2021/10/09 02:33:26 [INFO] [migrator@/work/tmp/dir1] get the current remote state
2021/10/09 02:33:27 [DEBUG] [executor@/work/tmp/dir1]$ terraform state pull
2021/10/09 02:33:27 [INFO] [migrator@/work/tmp/dir1] override backend to local
2021/10/09 02:33:27 [INFO] [executor@/work/tmp/dir1] create an override file
2021/10/09 02:33:27 [INFO] [migrator@/work/tmp/dir1] creating local workspace folder in: /work/tmp/dir1/terraform.tfstate.d/default
2021/10/09 02:33:27 [INFO] [executor@/work/tmp/dir1] switch backend to local
2021/10/09 02:33:29 [DEBUG] [executor@/work/tmp/dir1]$ terraform init -input=false -no-color -reconfigure
2021/10/09 02:33:29 [DEBUG] [executor@/work/tmp/dir2]$ terraform version
2021/10/09 02:33:29 [INFO] [migrator@/work/tmp/dir2] terraform version: 1.0.8
2021/10/09 02:33:29 [INFO] [migrator@/work/tmp/dir2] initialize work dir
2021/10/09 02:33:30 [DEBUG] [executor@/work/tmp/dir2]$ terraform init -input=false -no-color
2021/10/09 02:33:30 [INFO] [migrator@/work/tmp/dir2] switch to remote workspace default
2021/10/09 02:33:30 [DEBUG] [executor@/work/tmp/dir2]$ terraform workspace select default
2021/10/09 02:33:30 [INFO] [migrator@/work/tmp/dir2] get the current remote state
2021/10/09 02:33:31 [DEBUG] [executor@/work/tmp/dir2]$ terraform state pull
2021/10/09 02:33:31 [INFO] [migrator@/work/tmp/dir2] override backend to local
2021/10/09 02:33:31 [INFO] [executor@/work/tmp/dir2] create an override file
2021/10/09 02:33:31 [INFO] [migrator@/work/tmp/dir2] creating local workspace folder in: /work/tmp/dir2/terraform.tfstate.d/default
2021/10/09 02:33:31 [INFO] [executor@/work/tmp/dir2] switch backend to local
2021/10/09 02:33:33 [DEBUG] [executor@/work/tmp/dir2]$ terraform init -input=false -no-color -reconfigure
2021/10/09 02:33:33 [INFO] [migrator] compute new states (/work/tmp/dir1 => /work/tmp/dir2)
2021/10/09 02:33:33 [DEBUG] [executor@/work/tmp/dir1]$ terraform state mv -state=/tmp/tmp367964524 -state-out=/tmp/tmp726003163 -backup=/dev/null aws_security_group.foo aws_security_group.foo
2021/10/09 02:33:33 [DEBUG] [executor@/work/tmp/dir2]$ terraform state mv -state=/tmp/tmp624558974 -state-out=/tmp/tmp040605381 -backup=/dev/null aws_security_group.foo aws_security_group.foo2
2021/10/09 02:33:33 [DEBUG] [executor@/work/tmp/dir1]$ terraform state mv -state=/tmp/tmp080343392 -state-out=/tmp/tmp949845311 -backup=/dev/null aws_security_group.bar aws_security_group.bar
2021/10/09 02:33:33 [DEBUG] [executor@/work/tmp/dir2]$ terraform state mv -state=/tmp/tmp394029458 -state-out=/tmp/tmp326611401 -backup=/dev/null aws_security_group.bar aws_security_group.bar2
2021/10/09 02:33:33 [INFO] [migrator@/work/tmp/dir1] check diffs
2021/10/09 02:33:34 [DEBUG] [executor@/work/tmp/dir1]$ terraform plan -state=/tmp/tmp025293972 -out=/tmp/tfplan106779363 -input=false -no-color -detailed-exitcode
2021/10/09 02:33:34 [INFO] [migrator@/work/tmp/dir2] check diffs
2021/10/09 02:33:36 [DEBUG] [executor@/work/tmp/dir2]$ terraform plan -state=/tmp/tmp135992038 -out=/tmp/tfplan039088397 -input=false -no-color -detailed-exitcode
2021/10/09 02:33:36 [DEBUG] [executor@/work/tmp/dir2] failed to run command: (*exec.ExitError)(0xc00000e3e0)(exit status 2)
2021/10/09 02:33:36 [ERROR] [migrator@/work/tmp/dir2] unexpected diffs
2021/10/09 02:33:36 [INFO] [executor@/work/tmp/dir2] remove the override file
2021/10/09 02:33:36 [INFO] [executor@/work/tmp/dir2] remove the workspace state folder
2021/10/09 02:33:36 [INFO] [executor@/work/tmp/dir2] switch back to remote
2021/10/09 02:33:38 [DEBUG] [executor@/work/tmp/dir2]$ terraform init -input=false -no-color -reconfigure
2021/10/09 02:33:38 [INFO] [executor@/work/tmp/dir1] remove the override file
2021/10/09 02:33:38 [INFO] [executor@/work/tmp/dir1] remove the workspace state folder
2021/10/09 02:33:39 [INFO] [executor@/work/tmp/dir1] switch back to remote
2021/10/09 02:33:41 [DEBUG] [executor@/work/tmp/dir1]$ terraform init -input=false -no-color -reconfigure
terraform plan command returns unexpected diffs: failed to run command (exited 2): terraform plan -state=/tmp/tmp135992038 -out=/tmp/tfplan039088397 -input=false -no-color -detailed-exitcode
stdout:
aws_security_group.bar2: Refreshing state... [id=sg-3e592b7f]
aws_security_group.foo2: Refreshing state... [id=sg-27e71e85]
aws_security_group.bar: Refreshing state... [id=sg-3e592b7f]
aws_security_group.foo: Refreshing state... [id=sg-27e71e85]

Terraform used the selected providers to generate the following execution
plan. Resource actions are indicated with the following symbols:
  - destroy

Terraform will perform the following actions:

  # aws_security_group.bar2 will be destroyed
  - resource "aws_security_group" "bar2" {
      - arn                    = "arn:aws:ec2:ap-northeast-1:000000000000:security-group/sg-3e592b7f" -> null
      - description            = "Managed by Terraform" -> null
      - egress                 = [
          - {
              - cidr_blocks      = [
                  - "0.0.0.0/0",
                ]
              - description      = ""
              - from_port        = 0
              - ipv6_cidr_blocks = []
              - prefix_list_ids  = []
              - protocol         = "-1"
              - security_groups  = []
              - self             = false
              - to_port          = 0
            },
        ] -> null
      - id                     = "sg-3e592b7f" -> null
      - ingress                = [] -> null
      - name                   = "bar" -> null
      - owner_id               = "000000000000" -> null
      - revoke_rules_on_delete = false -> null
      - tags                   = {} -> null
      - tags_all               = {} -> null
    }

  # aws_security_group.foo2 will be destroyed
  - resource "aws_security_group" "foo2" {
      - arn                    = "arn:aws:ec2:ap-northeast-1:000000000000:security-group/sg-27e71e85" -> null
      - description            = "Managed by Terraform" -> null
      - egress                 = [
          - {
              - cidr_blocks      = [
                  - "0.0.0.0/0",
                ]
              - description      = ""
              - from_port        = 0
              - ipv6_cidr_blocks = []
              - prefix_list_ids  = []
              - protocol         = "-1"
              - security_groups  = []
              - self             = false
              - to_port          = 0
            },
        ] -> null
      - id                     = "sg-27e71e85" -> null
      - ingress                = [] -> null
      - name                   = "foo" -> null
      - owner_id               = "000000000000" -> null
      - revoke_rules_on_delete = false -> null
      - tags                   = {} -> null
      - tags_all               = {} -> null
    }

Plan: 0 to add, 0 to change, 2 to destroy.

─────────────────────────────────────────────────────────────────────────────

Saved the plan to: /tmp/tfplan039088397

To perform exactly these actions, run the following command to apply:
    terraform apply "/tmp/tfplan039088397"

stderr:
```

Am I doing this correctly this time? or should I have main.tf in both dir1 and dir2?

remove whole repo and re-clone from start, would you mind take a look for me once again when you have a chance? thanks! @minamijoyo 

```
bash-5.0# terraform state list
aws_security_group.bar
aws_security_group.foo
```

```
bash-5.0# pwd
/work/tmp/dir1
bash-5.0# cat main.tf 
NULL

bash-5.0# cat tfmigrate_mv_dir1_dir2.hcl 
migration "multi_state" "mv_dir1_dir2" {
  from_dir = "/work/tmp/dir1"
  to_dir   = "/work/tmp/dir2"
  actions = [
    "mv aws_security_group.foo aws_security_group.foo2",
    "mv aws_security_group.bar aws_security_group.bar2",
  ]
}
```

```
bash-5.0# cat ../dir2/main.tf 
resource "aws_security_group" "foo2" {
  name = "foo"
}

resource "aws_security_group" "bar2" {
  name = "bar"
}
```

```
bash-5.0# TFMIGRATE_LOG=DEBUG tfmigrate plan tfmigrate_mv_dir1_dir2.hcl 
2021/10/09 03:21:10 [DEBUG] [main] start: tfmigrate plan tfmigrate_mv_dir1_dir2.hcl
2021/10/09 03:21:10 [DEBUG] [main] tfmigrate version: 0.2.9
2021/10/09 03:21:10 [DEBUG] [command] config: &config.TfmigrateConfig{MigrationDir:".", History:(*history.Config)(nil)}
2021/10/09 03:21:10 [DEBUG] [command] option: &tfmigrate.MigratorOption{ExecPath:"", PlanOut:""}
2021/10/09 03:21:10 [INFO] [runner] load migration file: tfmigrate_mv_dir1_dir2.hcl
2021/10/09 03:21:10 [INFO] [migrator] multi start state migrator plan
2021/10/09 03:21:10 [DEBUG] [executor@/work/tmp/dir1]$ terraform version
2021/10/09 03:21:10 [INFO] [migrator@/work/tmp/dir1] terraform version: 1.0.8
2021/10/09 03:21:10 [INFO] [migrator@/work/tmp/dir1] initialize work dir
2021/10/09 03:21:11 [DEBUG] [executor@/work/tmp/dir1]$ terraform init -input=false -no-color
2021/10/09 03:21:11 [INFO] [migrator@/work/tmp/dir1] switch to remote workspace default
2021/10/09 03:21:12 [DEBUG] [executor@/work/tmp/dir1]$ terraform workspace select default
2021/10/09 03:21:12 [INFO] [migrator@/work/tmp/dir1] get the current remote state
2021/10/09 03:21:12 [DEBUG] [executor@/work/tmp/dir1]$ terraform state pull
2021/10/09 03:21:12 [INFO] [migrator@/work/tmp/dir1] override backend to local
2021/10/09 03:21:12 [INFO] [executor@/work/tmp/dir1] create an override file
2021/10/09 03:21:12 [INFO] [migrator@/work/tmp/dir1] creating local workspace folder in: /work/tmp/dir1/terraform.tfstate.d/default
2021/10/09 03:21:12 [INFO] [executor@/work/tmp/dir1] switch backend to local
2021/10/09 03:21:14 [DEBUG] [executor@/work/tmp/dir1]$ terraform init -input=false -no-color -reconfigure
2021/10/09 03:21:14 [DEBUG] [executor@/work/tmp/dir2]$ terraform version
2021/10/09 03:21:14 [INFO] [migrator@/work/tmp/dir2] terraform version: 1.0.8
2021/10/09 03:21:14 [INFO] [migrator@/work/tmp/dir2] initialize work dir
2021/10/09 03:21:15 [DEBUG] [executor@/work/tmp/dir2]$ terraform init -input=false -no-color
2021/10/09 03:21:15 [INFO] [migrator@/work/tmp/dir2] switch to remote workspace default
2021/10/09 03:21:16 [DEBUG] [executor@/work/tmp/dir2]$ terraform workspace select default
2021/10/09 03:21:16 [INFO] [migrator@/work/tmp/dir2] get the current remote state
2021/10/09 03:21:16 [DEBUG] [executor@/work/tmp/dir2]$ terraform state pull
2021/10/09 03:21:16 [INFO] [migrator@/work/tmp/dir2] override backend to local
2021/10/09 03:21:16 [INFO] [executor@/work/tmp/dir2] create an override file
2021/10/09 03:21:16 [INFO] [migrator@/work/tmp/dir2] creating local workspace folder in: /work/tmp/dir2/terraform.tfstate.d/default
2021/10/09 03:21:16 [INFO] [executor@/work/tmp/dir2] switch backend to local
2021/10/09 03:21:18 [DEBUG] [executor@/work/tmp/dir2]$ terraform init -input=false -no-color -reconfigure
2021/10/09 03:21:18 [INFO] [migrator] compute new states (/work/tmp/dir1 => /work/tmp/dir2)
2021/10/09 03:21:18 [DEBUG] [executor@/work/tmp/dir1]$ terraform state mv -state=/tmp/tmp606534563 -state-out=/tmp/tmp764914854 -backup=/dev/null aws_security_group.foo aws_security_group.foo
2021/10/09 03:21:18 [DEBUG] [executor@/work/tmp/dir2]$ terraform state mv -state=/tmp/tmp626163149 -state-out=/tmp/tmp429340360 -backup=/dev/null aws_security_group.foo aws_security_group.foo2
2021/10/09 03:21:19 [DEBUG] [executor@/work/tmp/dir1]$ terraform state mv -state=/tmp/tmp247318407 -state-out=/tmp/tmp407996730 -backup=/dev/null aws_security_group.bar aws_security_group.bar
2021/10/09 03:21:19 [DEBUG] [executor@/work/tmp/dir2]$ terraform state mv -state=/tmp/tmp750598481 -state-out=/tmp/tmp517987452 -backup=/dev/null aws_security_group.bar aws_security_group.bar2
2021/10/09 03:21:19 [INFO] [migrator@/work/tmp/dir1] check diffs
2021/10/09 03:21:19 [DEBUG] [executor@/work/tmp/dir1]$ terraform plan -state=/tmp/tmp112175019 -out=/tmp/tfplan772214030 -input=false -no-color -detailed-exitcode
2021/10/09 03:21:19 [INFO] [migrator@/work/tmp/dir2] check diffs
2021/10/09 03:21:21 [DEBUG] [executor@/work/tmp/dir2]$ terraform plan -state=/tmp/tmp198648085 -out=/tmp/tfplan739372912 -input=false -no-color -detailed-exitcode
2021/10/09 03:21:21 [DEBUG] [executor@/work/tmp/dir2] failed to run command: (*exec.ExitError)(0xc0002d1080)(exit status 2)
2021/10/09 03:21:21 [ERROR] [migrator@/work/tmp/dir2] unexpected diffs
2021/10/09 03:21:21 [INFO] [executor@/work/tmp/dir2] remove the override file
2021/10/09 03:21:21 [INFO] [executor@/work/tmp/dir2] remove the workspace state folder
2021/10/09 03:21:21 [INFO] [executor@/work/tmp/dir2] switch back to remote
2021/10/09 03:21:24 [DEBUG] [executor@/work/tmp/dir2]$ terraform init -input=false -no-color -reconfigure
2021/10/09 03:21:24 [INFO] [executor@/work/tmp/dir1] remove the override file
2021/10/09 03:21:24 [INFO] [executor@/work/tmp/dir1] remove the workspace state folder
2021/10/09 03:21:24 [INFO] [executor@/work/tmp/dir1] switch back to remote
2021/10/09 03:21:26 [DEBUG] [executor@/work/tmp/dir1]$ terraform init -input=false -no-color -reconfigure
terraform plan command returns unexpected diffs: failed to run command (exited 2): terraform plan -state=/tmp/tmp198648085 -out=/tmp/tfplan739372912 -input=false -no-color -detailed-exitcode
stdout:
aws_security_group.foo: Refreshing state... [id=sg-8dea8701]
aws_security_group.bar2: Refreshing state... [id=sg-8a179b01]
aws_security_group.bar: Refreshing state... [id=sg-8a179b01]
aws_security_group.foo2: Refreshing state... [id=sg-8dea8701]

Note: Objects have changed outside of Terraform

Terraform detected the following changes made outside of Terraform since the
last "terraform apply":

  # aws_security_group.bar2 has been changed
  ~ resource "aws_security_group" "bar2" {
        id                     = "sg-8a179b01"
        name                   = "bar"
      + tags                   = {}
        # (7 unchanged attributes hidden)
    }
  # aws_security_group.foo has been changed
  ~ resource "aws_security_group" "foo" {
        id                     = "sg-8dea8701"
        name                   = "foo"
      + tags                   = {}
        # (7 unchanged attributes hidden)
    }
  # aws_security_group.foo2 has been changed
  ~ resource "aws_security_group" "foo2" {
        id                     = "sg-8dea8701"
        name                   = "foo"
      + tags                   = {}
        # (7 unchanged attributes hidden)
    }
  # aws_security_group.bar has been changed
  ~ resource "aws_security_group" "bar" {
        id                     = "sg-8a179b01"
        name                   = "bar"
      + tags                   = {}
        # (7 unchanged attributes hidden)
    }

Unless you have made equivalent changes to your configuration, or ignored the
relevant attributes using ignore_changes, the following plan may include
actions to undo or respond to these changes.

─────────────────────────────────────────────────────────────────────────────

Terraform used the selected providers to generate the following execution
plan. Resource actions are indicated with the following symbols:
  - destroy

Terraform will perform the following actions:

  # aws_security_group.bar will be destroyed
  - resource "aws_security_group" "bar" {
      - arn                    = "arn:aws:ec2:ap-northeast-1:000000000000:security-group/sg-8a179b01" -> null
      - description            = "Managed by Terraform" -> null
      - egress                 = [
          - {
              - cidr_blocks      = [
                  - "0.0.0.0/0",
                ]
              - description      = ""
              - from_port        = 0
              - ipv6_cidr_blocks = []
              - prefix_list_ids  = []
              - protocol         = "-1"
              - security_groups  = []
              - self             = false
              - to_port          = 0
            },
        ] -> null
      - id                     = "sg-8a179b01" -> null
      - ingress                = [] -> null
      - name                   = "bar" -> null
      - owner_id               = "000000000000" -> null
      - revoke_rules_on_delete = false -> null
      - tags                   = {} -> null
      - tags_all               = {} -> null
    }

  # aws_security_group.foo will be destroyed
  - resource "aws_security_group" "foo" {
      - arn                    = "arn:aws:ec2:ap-northeast-1:000000000000:security-group/sg-8dea8701" -> null
      - description            = "Managed by Terraform" -> null
      - egress                 = [
          - {
              - cidr_blocks      = [
                  - "0.0.0.0/0",
                ]
              - description      = ""
              - from_port        = 0
              - ipv6_cidr_blocks = []
              - prefix_list_ids  = []
              - protocol         = "-1"
              - security_groups  = []
              - self             = false
              - to_port          = 0
            },
        ] -> null
      - id                     = "sg-8dea8701" -> null
      - ingress                = [] -> null
      - name                   = "foo" -> null
      - owner_id               = "000000000000" -> null
      - revoke_rules_on_delete = false -> null
      - tags                   = {} -> null
      - tags_all               = {} -> null
    }

Plan: 0 to add, 0 to change, 2 to destroy.

─────────────────────────────────────────────────────────────────────────────

Saved the plan to: /tmp/tfplan739372912

To perform exactly these actions, run the following command to apply:
    terraform apply "/tmp/tfplan739372912"

stderr:

```

yeah that is what I think therefore I raise this issue.

for assumption 1:
```
bash-5.0# cd ../dir2
bash-5.0# ls -al
total 12
drwxr-xr-x    6 root     root           192 Oct  9 03:44 .
drwxr-xr-x    4 root     root           128 Oct  9 03:17 ..
drwxr-xr-x    4 root     root           128 Oct  9 03:44 .terraform
-rw-r--r--    1 root     root           252 Oct  9 03:21 .terraform.lock.hcl
-rw-r--r--    1 root     root          1114 Oct  9 03:19 config.tf
-rw-r--r--    1 root     root           113 Oct  9 03:17 main.tf
bash-5.0# rm -rf .terraform*
bash-5.0# ls -al
total 8
drwxr-xr-x    4 root     root           128 Oct  9 04:11 .
drwxr-xr-x    4 root     root           128 Oct  9 03:17 ..
-rw-r--r--    1 root     root          1114 Oct  9 03:19 config.tf
-rw-r--r--    1 root     root           113 Oct  9 03:17 main.tf
```

```
bash-5.0# TFMIGRATE_LOG=DEBUG tfmigrate plan tfmigrate_mv_dir1_dir2.hcl 

2021/10/09 04:11:56 [DEBUG] [main] start: tfmigrate plan tfmigrate_mv_dir1_dir2.hcl
2021/10/09 04:11:56 [DEBUG] [main] tfmigrate version: 0.2.9
2021/10/09 04:11:56 [DEBUG] [command] config: &config.TfmigrateConfig{MigrationDir:".", History:(*history.Config)(nil)}
2021/10/09 04:11:56 [DEBUG] [command] option: &tfmigrate.MigratorOption{ExecPath:"", PlanOut:""}
2021/10/09 04:11:56 [INFO] [runner] load migration file: tfmigrate_mv_dir1_dir2.hcl
2021/10/09 04:11:56 [INFO] [migrator] multi start state migrator plan
2021/10/09 04:11:56 [DEBUG] [executor@/work/tmp/dir1]$ terraform version
2021/10/09 04:11:56 [INFO] [migrator@/work/tmp/dir1] terraform version: 1.0.8
2021/10/09 04:11:56 [INFO] [migrator@/work/tmp/dir1] initialize work dir
2021/10/09 04:11:58 [DEBUG] [executor@/work/tmp/dir1]$ terraform init -input=false -no-color
2021/10/09 04:11:58 [INFO] [migrator@/work/tmp/dir1] switch to remote workspace default
2021/10/09 04:11:58 [DEBUG] [executor@/work/tmp/dir1]$ terraform workspace select default
2021/10/09 04:11:58 [INFO] [migrator@/work/tmp/dir1] get the current remote state
2021/10/09 04:11:59 [DEBUG] [executor@/work/tmp/dir1]$ terraform state pull
2021/10/09 04:11:59 [INFO] [migrator@/work/tmp/dir1] override backend to local
2021/10/09 04:11:59 [INFO] [executor@/work/tmp/dir1] create an override file
2021/10/09 04:11:59 [INFO] [migrator@/work/tmp/dir1] creating local workspace folder in: /work/tmp/dir1/terraform.tfstate.d/default
2021/10/09 04:11:59 [INFO] [executor@/work/tmp/dir1] switch backend to local
2021/10/09 04:12:01 [DEBUG] [executor@/work/tmp/dir1]$ terraform init -input=false -no-color -reconfigure
2021/10/09 04:12:01 [DEBUG] [executor@/work/tmp/dir2]$ terraform version
2021/10/09 04:12:01 [INFO] [migrator@/work/tmp/dir2] terraform version: 1.0.8
2021/10/09 04:12:01 [INFO] [migrator@/work/tmp/dir2] initialize work dir
2021/10/09 04:12:02 [DEBUG] [executor@/work/tmp/dir2]$ terraform init -input=false -no-color
2021/10/09 04:12:02 [INFO] [migrator@/work/tmp/dir2] switch to remote workspace default
2021/10/09 04:12:03 [DEBUG] [executor@/work/tmp/dir2]$ terraform workspace select default
2021/10/09 04:12:03 [INFO] [migrator@/work/tmp/dir2] get the current remote state
2021/10/09 04:12:03 [DEBUG] [executor@/work/tmp/dir2]$ terraform state pull
2021/10/09 04:12:03 [INFO] [migrator@/work/tmp/dir2] override backend to local
2021/10/09 04:12:03 [INFO] [executor@/work/tmp/dir2] create an override file
2021/10/09 04:12:03 [INFO] [migrator@/work/tmp/dir2] creating local workspace folder in: /work/tmp/dir2/terraform.tfstate.d/default
2021/10/09 04:12:03 [INFO] [executor@/work/tmp/dir2] switch backend to local
2021/10/09 04:12:05 [DEBUG] [executor@/work/tmp/dir2]$ terraform init -input=false -no-color -reconfigure
2021/10/09 04:12:05 [INFO] [migrator] compute new states (/work/tmp/dir1 => /work/tmp/dir2)
2021/10/09 04:12:05 [DEBUG] [executor@/work/tmp/dir1]$ terraform state mv -state=/tmp/tmp518146889 -state-out=/tmp/tmp750608404 -backup=/dev/null aws_security_group.foo aws_security_group.foo
2021/10/09 04:12:05 [DEBUG] [executor@/work/tmp/dir2]$ terraform state mv -state=/tmp/tmp481165923 -state-out=/tmp/tmp939930214 -backup=/dev/null aws_security_group.foo aws_security_group.foo2
2021/10/09 04:12:05 [DEBUG] [executor@/work/tmp/dir1]$ terraform state mv -state=/tmp/tmp400547981 -state-out=/tmp/tmp023251592 -backup=/dev/null aws_security_group.bar aws_security_group.bar
2021/10/09 04:12:05 [DEBUG] [executor@/work/tmp/dir2]$ terraform state mv -state=/tmp/tmp919773767 -state-out=/tmp/tmp940621050 -backup=/dev/null aws_security_group.bar aws_security_group.bar2
2021/10/09 04:12:05 [INFO] [migrator@/work/tmp/dir1] check diffs
2021/10/09 04:12:06 [DEBUG] [executor@/work/tmp/dir1]$ terraform plan -state=/tmp/tmp889071633 -out=/tmp/tfplan452379196 -input=false -no-color -detailed-exitcode
2021/10/09 04:12:06 [INFO] [migrator@/work/tmp/dir2] check diffs
2021/10/09 04:12:08 [DEBUG] [executor@/work/tmp/dir2]$ terraform plan -state=/tmp/tmp828791403 -out=/tmp/tfplan737042126 -input=false -no-color -detailed-exitcode
2021/10/09 04:12:08 [DEBUG] [executor@/work/tmp/dir2] failed to run command: (*exec.ExitError)(0xc00040e320)(exit status 2)
2021/10/09 04:12:08 [ERROR] [migrator@/work/tmp/dir2] unexpected diffs
2021/10/09 04:12:08 [INFO] [executor@/work/tmp/dir2] remove the override file
2021/10/09 04:12:08 [INFO] [executor@/work/tmp/dir2] remove the workspace state folder
2021/10/09 04:12:08 [INFO] [executor@/work/tmp/dir2] switch back to remote
2021/10/09 04:12:10 [DEBUG] [executor@/work/tmp/dir2]$ terraform init -input=false -no-color -reconfigure
2021/10/09 04:12:10 [INFO] [executor@/work/tmp/dir1] remove the override file
2021/10/09 04:12:10 [INFO] [executor@/work/tmp/dir1] remove the workspace state folder
2021/10/09 04:12:10 [INFO] [executor@/work/tmp/dir1] switch back to remote
2021/10/09 04:12:12 [DEBUG] [executor@/work/tmp/dir1]$ terraform init -input=false -no-color -reconfigure
terraform plan command returns unexpected diffs: failed to run command (exited 2): terraform plan -state=/tmp/tmp828791403 -out=/tmp/tfplan737042126 -input=false -no-color -detailed-exitcode
stdout:
aws_security_group.foo: Refreshing state... [id=sg-8dea8701]
aws_security_group.bar: Refreshing state... [id=sg-8a179b01]
aws_security_group.foo2: Refreshing state... [id=sg-8dea8701]
aws_security_group.bar2: Refreshing state... [id=sg-8a179b01]

Note: Objects have changed outside of Terraform

Terraform detected the following changes made outside of Terraform since the
last "terraform apply":

  # aws_security_group.foo2 has been changed
  ~ resource "aws_security_group" "foo2" {
        id                     = "sg-8dea8701"
        name                   = "foo"
      + tags                   = {}
        # (7 unchanged attributes hidden)
    }
  # aws_security_group.bar has been changed
  ~ resource "aws_security_group" "bar" {
        id                     = "sg-8a179b01"
        name                   = "bar"
      + tags                   = {}
        # (7 unchanged attributes hidden)
    }
  # aws_security_group.bar2 has been changed
  ~ resource "aws_security_group" "bar2" {
        id                     = "sg-8a179b01"
        name                   = "bar"
      + tags                   = {}
        # (7 unchanged attributes hidden)
    }
  # aws_security_group.foo has been changed
  ~ resource "aws_security_group" "foo" {
        id                     = "sg-8dea8701"
        name                   = "foo"
      + tags                   = {}
        # (7 unchanged attributes hidden)
    }

Unless you have made equivalent changes to your configuration, or ignored the
relevant attributes using ignore_changes, the following plan may include
actions to undo or respond to these changes.

─────────────────────────────────────────────────────────────────────────────

Terraform used the selected providers to generate the following execution
plan. Resource actions are indicated with the following symbols:
  - destroy

Terraform will perform the following actions:

  # aws_security_group.bar will be destroyed
  - resource "aws_security_group" "bar" {
      - arn                    = "arn:aws:ec2:ap-northeast-1:000000000000:security-group/sg-8a179b01" -> null
      - description            = "Managed by Terraform" -> null
      - egress                 = [
          - {
              - cidr_blocks      = [
                  - "0.0.0.0/0",
                ]
              - description      = ""
              - from_port        = 0
              - ipv6_cidr_blocks = []
              - prefix_list_ids  = []
              - protocol         = "-1"
              - security_groups  = []
              - self             = false
              - to_port          = 0
            },
        ] -> null
      - id                     = "sg-8a179b01" -> null
      - ingress                = [] -> null
      - name                   = "bar" -> null
      - owner_id               = "000000000000" -> null
      - revoke_rules_on_delete = false -> null
      - tags                   = {} -> null
      - tags_all               = {} -> null
    }

  # aws_security_group.foo will be destroyed
  - resource "aws_security_group" "foo" {
      - arn                    = "arn:aws:ec2:ap-northeast-1:000000000000:security-group/sg-8dea8701" -> null
      - description            = "Managed by Terraform" -> null
      - egress                 = [
          - {
              - cidr_blocks      = [
                  - "0.0.0.0/0",
                ]
              - description      = ""
              - from_port        = 0
              - ipv6_cidr_blocks = []
              - prefix_list_ids  = []
              - protocol         = "-1"
              - security_groups  = []
              - self             = false
              - to_port          = 0
            },
        ] -> null
      - id                     = "sg-8dea8701" -> null
      - ingress                = [] -> null
      - name                   = "foo" -> null
      - owner_id               = "000000000000" -> null
      - revoke_rules_on_delete = false -> null
      - tags                   = {} -> null
      - tags_all               = {} -> null
    }

Plan: 0 to add, 0 to change, 2 to destroy.

─────────────────────────────────────────────────────────────────────────────

Saved the plan to: /tmp/tfplan737042126

To perform exactly these actions, run the following command to apply:
    terraform apply "/tmp/tfplan737042126"

stderr:

```

For assumption 2: (how can I confirm the state are not in the same location? the terraform.tfstate file looks strange to me I though they should have 2 entries)
```
bash-5.0# pwd
/work/tmp/dir1
bash-5.0# cat .terraform/terraform.tfstate 
{
    "version": 3,
    "serial": 7,
    "lineage": "e2df0afd-129a-3422-0223-0ad8a990e64c",
    "backend": {
        "type": "s3",
        "config": {
            "access_key": "dummy",
            "acl": null,
            "assume_role_duration_seconds": null,
            "assume_role_policy": null,
            "assume_role_policy_arns": null,
            "assume_role_tags": null,
            "assume_role_transitive_tag_keys": null,
            "bucket": "tfstate-test",
            "dynamodb_endpoint": null,
            "dynamodb_table": null,
            "encrypt": null,
            "endpoint": "http://localstack:4566",
            "external_id": null,
            "force_path_style": true,
            "iam_endpoint": null,
            "key": "test/terraform.tfstate",
            "kms_key_id": null,
            "max_retries": null,
            "profile": null,
            "region": "ap-northeast-1",
            "role_arn": null,
            "secret_key": "dummy",
            "session_name": null,
            "shared_credentials_file": null,
            "skip_credentials_validation": true,
            "skip_metadata_api_check": true,
            "skip_region_validation": null,
            "sse_customer_key": null,
            "sts_endpoint": null,
            "token": null,
            "workspace_key_prefix": null
        },
        "hash": 3459050205
    },
    "modules": [
        {
            "path": [
                "root"
            ],
            "outputs": {},
            "resources": {},
            "depends_on": []
        }
    ]
}
```

and 
```
bash-5.0# pwd
/work/tmp/dir2
bash-5.0# cat .terraform/terraform.tfstate 
{
    "version": 3,
    "serial": 3,
    "lineage": "d17f3ead-0684-9fd5-dcd3-847da8e54812",
    "backend": {
        "type": "s3",
        "config": {
            "access_key": "dummy",
            "acl": null,
            "assume_role_duration_seconds": null,
            "assume_role_policy": null,
            "assume_role_policy_arns": null,
            "assume_role_tags": null,
            "assume_role_transitive_tag_keys": null,
            "bucket": "tfstate-test",
            "dynamodb_endpoint": null,
            "dynamodb_table": null,
            "encrypt": null,
            "endpoint": "http://localstack:4566",
            "external_id": null,
            "force_path_style": true,
            "iam_endpoint": null,
            "key": "test/terraform.tfstate",
            "kms_key_id": null,
            "max_retries": null,
            "profile": null,
            "region": "ap-northeast-1",
            "role_arn": null,
            "secret_key": "dummy",
            "session_name": null,
            "shared_credentials_file": null,
            "skip_credentials_validation": true,
            "skip_metadata_api_check": true,
            "skip_region_validation": null,
            "sse_customer_key": null,
            "sts_endpoint": null,
            "token": null,
            "workspace_key_prefix": null
        },
        "hash": 3459050205
    },
    "modules": [
        {
            "path": [
                "root"
            ],
            "outputs": {},
            "resources": {},
            "depends_on": []
        }
    ]
}

```

--- a/README.md
+++ b/README.md
@@ -189,15 +189,74 @@ YYYY/MM/DD hh:mm:ss [INFO] [migrator] state migrator apply success!
 The apply command computes a new state and pushes it to remote state.
 It will fail if terraform plan detects any diffs with the new state.
 
-Finally, you can check the latest remote state has no changes with terraform plan:
+You can confirm the latest remote state has no changes with terraform plan:
 
 ```
 # terraform plan
 (snip.)
 No changes. Infrastructure is up-to-date.
+
+# terraform state list
+aws_security_group.bar
+aws_security_group.baz
 ```
 
 There is no magic. The tfmigrate just did the boring work for you.
+
+Furthermore, you can also move resources to another directory. Let's split the tfstate in two.
+Create a new empty directory with a different remote state path:
+
+```
+# mkdir dir2
+# cat config.tf | sed 's/test\/terraform.tfstate/dir2\/terraform.tfstate/' > dir2/config.tf
+```
+
+Move the resource definition of `aws_security_group.baz` in `main.tf` to `dir2/main.tf` and rename it to `aws_security_group.baz2`:
+
+```
+# cat << EOF > main.tf
+resource "aws_security_group" "bar" {
+  name = "bar"
+}
+EOF
+
+# cat << EOF > dir2/main.tf
+resource "aws_security_group" "baz2" {
+  name = "foo"
+}
+EOF
+```
+
+Create a `multi_state` migration file:
+
+```
+# cat << EOF > tfmigrate_multi_state_test.hcl
+migration "multi_state" "test" {
+  from_dir = "."
+  to_dir   = "dir2"
+
+  actions = [
+    "mv aws_security_group.baz aws_security_group.baz2",
+  ]
+}
+EOF
+```
+
+Run tfmigrate plan & apply:
+
+```
+# tfmigrate plan tfmigrate_multi_state_test.hcl
+# tfmigrate apply tfmigrate_multi_state_test.hcl
+```
+
+You can see the tfstate was split in two:
+
+```
+# terraform state list
+aws_security_group.bar
+# cd dir2 && terraform state list
+aws_security_group.baz2
+```
 
 ## Install
 

